packages/template-retail-react-app/app/utils/responsive-image.js

```diff
@@ -64,29 +88,90 @@ const mapWidthsToSizes = (widths) => {
+    const flat = Array.isArray(collector)
+    return densities.reduce(
+        (acc, entry) => {
+            if (!flat && Array.isArray(entry) && Array.isArray(entry.at(0))) {
+                acc.push(mapDensities(entry, []))
+            } else if (
+                Array.isArray(entry) &&
+                typeof entry.at(0) === 'number' &&
+                !Number.isNaN(entry.at(0))
+            ) {
+                const w = entry.at(0)
+                const f = entry.at(1)
+                const result = [w, typeof f === 'number' && !Number.isNaN(f) ? f : w]
+                acc.push(flat ? result : [result])
+            } else if (typeof entry === 'number' && !Number.isNaN(entry)) {
+                const result = [entry, entry]
+                acc.push(flat ? result : [result])
+            } else if (flat) {
+                acc.push(...defaultDensityFactors)
+            } else {
+                acc.push(defaultDensityFactors)
+            }
+            return acc
+        },
+        flat ? collector : []
+    )
+}
 
-        _widths = [..._widths, ...Array(amountToPad).fill(lastWidth)]
+function obtainDensities(densities) {
+    const cache = densitiesCache.get(themeBreakpoints)
+    if (cache?.has(densities)) {
+        return densitiesCache.get(themeBreakpoints).get(densities)
+    } else if (isObject(densities)) {
+        const _densities = mapDensities(padArray(breakpointMapAsArray(densities)))
+        cache?.set(densities, _densities)
+        return _densities
+    } else if (Array.isArray(densities)) {
+        const _densities = mapDensities(padArray(densities))
+        cache?.set(densities, _densities)
+        return _densities
     }
+    return effectiveDefaultDensities
+}
 
-    _widths = uniqueArray(convertToPxNumbers(_widths)).sort()
-
-    const srcSet = []
-    _widths.forEach((width) => {
-        srcSet.push(width)
-        srcSet.push(width * 2) // for devices with higher pixel density
-    })
+/**
+ * @param {Object} props
+ * @param {string} props.src
+ * @param {(number[]|string[]|Object)} props.widths
+ * @param {(number[]|[number, number][])} props.densities
+ * @return {string}
+ */
+const mapWidthsToSrcSet = ({src, widths, densities}) => {
+    let _widths = isObject(widths) ? breakpointMapAsArray(widths) : widths.slice(0)
+    _widths = uniqueArray(convertToPxNumbers(padArray(_widths))).sort()
 
-    return srcSet.map((imageWidth) => `${getSrc(dynamicSrc, imageWidth)} ${imageWidth}w`).join(', ')
+    // Request images using given or default density factors. By default, images are requested with factors 1 and 2.
+    const _densities = obtainDensities(densities)
+    const set = _widths.reduce((acc, width, idx) => {
+        const densityFactors = _densities.at(idx)
+        for (const [w, f] of densityFactors) {
+            const w1 = Math.round(width * w)
+            const w2 = Math.round(width * f)
+            acc.add(`${getSrc(src, w2)} ${w1}w`)
```

Comment:
The width descriptor in srcSet (e.g., image-200.jpg 200w) must be the image width value  right? I think we want to use w2 and not w1.


E.g.:

If width = 100 and [w, f] = [1, 2], it generates: image-200.jpg 100w, which is wrong. 

```suggestion
            acc.add(`${getSrc(src, w2)} ${w2}w`)
```

Reply:
obsolete